### PR TITLE
[test] Add integration test for exposing ports

### PR DESCRIPTION
**Pull request by WVerlaek**
## Description

Tests the following:
- Port auto-detection
  - Workspace has a python server running at port 3000, which should get detected by supervisor
- Private port returning 401 Unauthorized when accessed from outside the workspace
- Setting port visibility to public through `gp ports visibility`
  - Also checks the port status is updated in the list of ports in supervisor
- Checks that a public port is accessible from outside the workspace

## Related Issue(s)

Fixes WKS-142

## How to test

Run workspace integration tests

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - wv-mk2-tesbcc5ee9b01</li>
	<li><b>🔗 URL</b> - <a href="https://wv-mk2-tesbcc5ee9b01.preview.gitpod-dev.com/workspaces" target="_blank">wv-mk2-tesbcc5ee9b01.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] with-integration-tests=workspace
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: test/tests/workspace/ports_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+				// nftable rule only forwards to this ip address
+				Args: []string{"ports", "visibility", "3000:public"},
+			}, &res1)
+			if err != nil {
+				t.Fatal(err)
+			}
+			t.Logf("debug: gp CLI output: %s, %s, %d", res1.Stdout, res1.Stderr, res1.ExitCode)
+
+			// Wait for workspace to pick up the port change.
+			time.Sleep(3 * time.Second)
```
[the remark this answers is not in the document]
Comment (WVerlaek) [reply test/tests/workspace/ports_test.go:182]: good point, I can add a retry mechanism to check for some time? Or were you thinking of a different way to subscribe to port changes?

[the remark this answers is not in the document]
Comment (WVerlaek) [reply test/tests/workspace/ports_test.go]: Added a retry

[the remark this answers is not in the document]
Comment (akosyakov) [reply test/tests/workspace/ports_test.go]: I was thinking about using supervisor stream API for ports

[the remark this answers is not in the document]
Comment (akosyakov) [reply test/tests/workspace/ports_test.go]: like in https://github.com/gitpod-io/gitpod/blob/bf947d1260d64a54f8f16c49f3c80f1332c1c381/components/gitpod-cli/pkg/supervisor/ports.go#L14 but with observe: true

[the remark this answers is not in the document]
Comment (akosyakov) [reply test/tests/workspace/ports_test.go]: although you will need retry as well if stream errors

[the remark this answers is not in the document]
Comment (WVerlaek) [reply test/tests/workspace/ports_test.go]: getting `rpc error: code = Unavailable desc = connection error: desc = "error reading server preface: http2: frame too large"` when trying the following:
```
client, err := api.Supervisor(instanceId)
if err != nil {
	t.Fatal(err)
}
statusClient := supervisor.NewStatusServiceClient(client)

// rpc error is returned here:
status, err := statusClient.PortsStatus(ctx, &supervisor.PortsStatusRequest{
	Observe: true,
})
```

have you seen this before? I don't see any other integration test using the `api.Supervisor` component so not sure if it's expected to work?

[the remark this answers is not in the document]
Comment (akosyakov) [reply test/tests/workspace/ports_test.go]: No, I have not seen i tbefore, maybe address is wrong, i would need to check. I think you previous attempt with retry is good.

**Review by Furisto — APPROVED** [review]

File: test/tests/workspace/ports_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (WVerlaek) [reply test/tests/workspace/ports_test.go]: @akosyakov any other feedback? Or good to merge? 🙏

**Review by akosyakov — APPROVED** [review]
sorry, it slipped

**Pull request merged**